git/tags.go

```diff
@@ -3,20 +3,96 @@ package git
 import (
 	"errors"
 	"fmt"
+	"strings"
 	"github.com/Masterminds/semver"
 	"github.com/go-git/go-git/v5"
 	"github.com/go-git/go-git/v5/plumbing"
 )
 
 type Tags = map[plumbing.Hash]*semver.Version
 
+// tagCandidate holds a tag candidate with its original name and parsed version
+type tagCandidate struct {
+	originalName string
+	version      *semver.Version
+}
+
+// getTagSpecificity returns how specific a tag is based on the number of version components.
+// More dots means more specific: "v4" (0 dots) < "v4.5" (1 dot) < "v4.5.14" (2 dots)
+func getTagSpecificity(tagName string) int {
+	// Remove 'v' prefix if present
+	cleanTag := tagName
+	if strings.HasPrefix(cleanTag, "v") {
+		cleanTag = cleanTag[1:]
+	}
+	return strings.Count(cleanTag, ".")
+}
+
+// areCompatibleGranularities checks if two versions are compatible granularities of the same version.
+// For example, v4.0.0, v4.5.0, and v4.5.14 are compatible if they represent granularities like v4, v4.5, v4.5.14
+func areCompatibleGranularities(v1, v2 *semver.Version) bool {
```

Comment:
Maybe we should not call the variables `v1` and `v2`, as this could be mistaken literally. Maybe rather use `leftVersion` and `rightVersion`. This not only applies here, but at other places as well.